jbpm-task-assigning/jbpm-task-assigning-core/src/test/java/org/jbpm/task/assigning/model/solver/realtime/AssignTaskProblemFactChangeTest.java

```diff
@@ -0,0 +1,320 @@
+        TaskAssigningSolution initialSolution = executeSequentialChanges(solution, programmedChanges);
+        if (writeTestFiles()) {
+            String resourceName = solutionResource.substring(solutionResource.lastIndexOf("/") + 1);
+            writeToTempFile("AssignTaskProblemFactChangeTest.assignTaskProblemFactChangeTest." + testType + ".InitialSolution_", printSolution(initialSolution));
+            for (int i = 0; i < programmedChanges.size(); i++) {
+                ProgrammedAssignTaskProblemFactChange scheduledChange = programmedChanges.get(i);
+                try {
+                    writeToTempFile("AssignTaskProblemFactChangeTest.assignTaskProblemFactChangeTest." + testType + ".WorkingSolutionBeforeChange_" + resourceName + "_" + i + "__", scheduledChange.workingSolutionBeforeChangeAsString());
+                    writeToTempFile("AssignTaskProblemFactChangeTest.assignTaskProblemFactChangeTest." + testType + ".SolutionAfterChange_" + resourceName + "_" + i + "__", scheduledChange.solutionAfterChangeAsString());
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+
+        //each partial solution must have the change that was applied on it.
+        programmedChanges.forEach(change -> assertAssignTaskProblemFactChangeWasProduced(change.getChange(), change.getSolutionAfterChange()));
+
+        //finally the last solution must have the result of all the changes.
+        TaskAssigningSolution lastSolution = programmedChanges.get(programmedChanges.size() - 1).getSolutionAfterChange();
+        Map<Long, AssignTaskProblemFactChange> summarizedChanges = new HashMap<>();
+        programmedChanges.forEach(change -> {
+            //if  task was changed multiple times record only the last change.
+            summarizedChanges.put(change.getChange().getTask().getId(), change.getChange());
+        });
+        summarizedChanges.values().forEach(change -> assertAssignTaskProblemFactChangeWasProduced(change, lastSolution));
+    }
+
+    /**
+     * Given an AssignTaskProblemFactChange and a solution that was produced as the result of applying the change,
+     * asserts that the assignment defined by the change is not violated (exists in) by the solution.
+     * The assignment defined in the change must also be pinned in the produced solution as well as any other
+     * previous assignment for the given user.
+     * @param change The change that was executed for producing the solution.
+     * @param solution The produced solution.
+     */
+    private void assertAssignTaskProblemFactChangeWasProduced(AssignTaskProblemFactChange change, TaskAssigningSolution solution) {
+        User internalUser = solution.getUserList().stream()
+                .filter(user -> Objects.equals(user.getId(), change.getUser().getId()))
+                .findFirst().orElse(null);
```

Comment:
Instead of asserting nullity of the returned value one line below, we can throw exception straight away.

Reply:
I agree, another valid option could be to throw an Exception. But for this case I like more the current one.

Reply:
Well, as on the line below, there is junit's `assertNotNull`, the only thing we learn when the `internalUser` is null will be information that the line 292 failed as it assumed the object be not null.
On the other hand, if we throw an exception from the stream directly, we can provide additional information that would help with investigating the failure.